### DISCUSSION TITLE
What are the differences in the query methods local, global, and question generation?

### DISCUSSION THREAD

**Discussion by natoverse** [microsoft/graphrag#721] [Algorithm + Paper]
Copied from #695 

# Question text
1. What is the relationship between the question generation module and global/local query module? I can't correlate them at the moment.
2. How is retrieval and re-reranking implemented for local queries? I understand retrieval by pre-computed embedding and reranking by calling LLM?

# Answer
The [paper](https://arxiv.org/pdf/2404.16130) has detailed descriptions of the global query algorithm, but I'll summarize briefly here along with local search:
- **Local search** uses entities as the entry point for building the context. All entity descriptions are embedded during indexing, and when a local search is executed we embed the query text to find the nearest entities. We can then use graph traversal to build the prompt context from the entities, relationships, and communities that are retrieved this way. In general, this makes local search better for targeted queries about entities.
- **Global search** uses the pre-computed summaries of each entity community to generate an answer. It does this by asking the question to every community, and then rolling these up into an overall summary response. In general, this makes global search better for broad thematic questions that don't reference and entity directly.

There's a bit more to it in terms of ranking matches, managing context window size limits, etc., which can be found in the paper and [code](https://github.com/microsoft/graphrag/blob/main/graphrag/query/input/retrieval/relationships.py).

**Question generation** may be helpful for use cases where you would like the model to provide starting points for analysis. This is useful because the model has access to all of the indexed content and can summarize it quickly compared to human reading. Question generation can also take prior user questions as input and suggest relevant next questions to ask. As a starting point for building the context in the absence of any prior conversation history, question generation [will retrieve](https://github.com/microsoft/graphrag/blob/main/graphrag/query/context_builder/entity_extraction.py) the top _k_ entities based on rank (rank is in turn computed as the node degree, so this effectively retrieves the top _k_ most connected entities). In this sense it functions much like local search because it uses entities as the entry point.

**Comment by KylinMountain** [comment]
so i have question why we just one local question gen in code if we have input history? Can we make it to global question？ @natoverse

**Reply by natoverse** [reply]
The local search comparison is only if you supply no history - i.e., we need to start with something, so we start with the top ranked entities since they are likely to carry a lot of weight in the dataset. If you supply your own history you can give it whatever context you want to frame the question. I think that answers the technical part of your question, but are you also looking for some default that makes it act more like global search?

**Comment by LIHUA919** [comment]
why

**Comment by paolo-tamagnini-tr** [comment]
can someone here confirm or deny whether the input entities of the local search can be tampered with? 
that is whether i can provide in my app a filter which filters the dataframe with the entities before the query runs? 
that would be helpful to make sure the answer will only be about selected entities.
